pkg/docker/config/config.go

```diff
@@ -0,0 +1,281 @@
+
+// DockerHostNameToRegistry converts the docker hostname to the docker registry
+func DockerHostNameToRegistry(registry string) string {
+	if registry == dockerHostname {
+		return dockerRegistry
+	}
+	return registry
+}
+
+// getPath gets the path of the auth.json file since the user can set their own path
+// by setting CRIO_AUTH_PATH
+// If CRIO_AUTH_PATH is not set, the default path is ~/.config/containers/auth.json
+func getPathToAuth() (string, error) {
+	if os.Getenv("CRIO_AUTH_PATH") != "" {
+		return filepath.Join(os.Getenv("CRIO_AUTH_PATH"), authCfgFileName), nil
+	}
+	dir := filepath.Join(homedir.Get(), authCfg)
+	if _, err := os.Stat(dir); os.IsNotExist(err) {
+		if err = os.Mkdir(dir, 0700); err != nil {
+			return "", errors.Wrapf(err, "error creating directory %q", dir)
+		}
+	}
+	return filepath.Join(dir, authCfgFileName), nil
+}
+
+// readJSONFile unmarshals the authentications stored in the auth.json file and returns it
+// or returns an empty dockerConfigFile data structure if auth.json is empty
+func readJSONFile(path string) (dockerConfigFile, error) {
+	var auths dockerConfigFile
+
+	raw, err := ioutil.ReadFile(path)
+	if os.IsNotExist(err) {
+		auths.AuthConfigs = map[string]dockerAuthConfig{}
+		return auths, nil
+	}
+
+	if err = json.Unmarshal(raw, &auths); err != nil {
+		return dockerConfigFile{}, errors.Wrapf(err, "error unmarshaling JSON at %q", path)
+	}
+
+	return auths, nil
+}
+
+// addAuthToJSON either adds a new registry url with its encoded user:pass
+// or modifies an existing registry url
+func modifyJSON(registry, username, password, path, action string) error {
+	auths, err := readJSONFile(path)
+	if err != nil {
+		return errors.Wrapf(err, "error reading JSON file %q", path)
+	}
+
+	switch action {
+	case addAuth:
+		if len(auths.CredHelpers) > 0 {
+			if err = setAuthToCredHelper(registry, username, password); err != nil {
+				return err
+			}
+			break
+		}
+
+		creds := base64.StdEncoding.EncodeToString([]byte(username + ":" + password))
+		newCreds := dockerAuthConfig{Auth: creds}
+		auths.AuthConfigs[registry] = newCreds
+	case deleteAuth:
+		// First try cred helpers.
+		if ch, exists := auths.CredHelpers[registry]; exists {
+			if err = deleteAuthFromCredHelper(ch, registry); err != nil {
+				return err
+			}
+			break
+		}
+
+		if _, ok := auths.AuthConfigs[registry]; ok {
+			delete(auths.AuthConfigs, registry)
+		} else if _, ok := auths.AuthConfigs[normalizeRegistry(registry)]; ok {
+			delete(auths.AuthConfigs, normalizeRegistry(registry))
+		} else {
+			return ErrNotLoggedIn
+		}
+	}
+
+	newData, err := json.MarshalIndent(auths, "", "\t")
+	if err != nil {
+		return errors.Wrapf(err, "error marshaling JSON %q", path)
+	}
+
+	if err = ioutil.WriteFile(path, newData, 0755); err != nil {
+		return errors.Wrapf(err, "error writing to file %q", path)
+	}
+	return nil
+}
+
+func getAuthFromCredHelper(credHelper, registry string) (string, string, error) {
+	registry = convertToHostname(registry)
```

Comment:
Is this necessary? I wouldn’t expect the callers of this package to use an URL for the `registry` parameter.

(Same in the other two cases.)

Reply:
fixed